File: untold_comparison_programs/07_security/javascript/main.js
// Security: Hashing & Encoding
const crypto = require('crypto');

const password = "MySecurePassword123";

// SHA256 hashing
const hash = crypto.createHash('sha256').update(password).digest('hex');
console.log(`SHA256: ${hash}`);

// SHA512 hashing
const hash512 = crypto.createHash('sha512').update(password).digest('hex');
console.log(`SHA512: ${hash512}`);

// HMAC
const hmacVal = crypto.createHmac('sha256', 'secret_key').update(password).digest('hex');
console.log(`HMAC-SHA256: ${hmacVal}`);

// Base64 encoding
const encoded = Buffer.from(password).toString('base64');
console.log(`Base64: ${encoded}`);

// Constant-time comparison
const storedHash = crypto.createHash('sha256').update('password').digest('hex');
if (crypto.timingSafeEqual(Buffer.from(storedHash), Buffer.from(hash))) {
    console.log("Secure match!");
}

// Verify hash
const computed = crypto.createHash('sha256').update(password).digest('hex');
if (hash === computed) {
    console.log("Hash verified!");
}

// Random token generation
const token = crypto.randomBytes(32).toString('hex');
console.log(`Random token: ${token}`);

File: untold_comparison_programs/07_security/python/main.py
# Security: Hashing & Encoding
import hashlib
import hmac
import base64
import secrets

password = "MySecurePassword123"

# SHA256 hashing
hash = hashlib.sha256(password.encode()).hexdigest()
print(f"SHA256: {hash}")

# SHA512 hashing
hash512 = hashlib.sha512(password.encode()).hexdigest()
print(f"SHA512: {hash512}")

# HMAC
hmac_val = hmac.new(b"secret_key", password.encode(), hashlib.sha256).hexdigest()
print(f"HMAC-SHA256: {hmac_val}")

# Base64 encoding
encoded = base64.b64encode(password.encode()).decode()
print(f"Base64: {encoded}")

# Constant-time comparison
stored_hash = hashlib.sha256(b"password").hexdigest()
if hmac.compare_digest(stored_hash, hash):
    print("Secure match!")

# Verify hash
computed = hashlib.sha256(password.encode()).hexdigest()
if hmac.compare_digest(computed, computed):
    print("Hash verified!")

# Random token generation
token = secrets.token_hex(32)
print(f"Random token: {token}")